rcl/src/rcl/event.c

```diff
@@ -0,0 +1,211 @@
+
+  // Allocate space for the implementation struct.
+  event->impl = (rcl_event_impl_t *) allocator->allocate(
+    sizeof(rcl_event_impl_t), allocator->state);
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    event->impl, "allocating memory failed", ret = RCL_RET_BAD_ALLOC; return ret);
+
+  rmw_event_type_t rmw_event_type;
+  switch (event_type) {
+    case RCL_SUBSCRIPTION_DEADLINE:
+      rmw_event_type = RMW_EVENT_REQUESTED_DEADLINE_MISSED;
+      break;
+    case RCL_SUBSCRIPTION_LIVELINESS:
+      rmw_event_type = RMW_EVENT_LIVELINESS_CHANGED;
+      break;
+    default:
+      return RCL_RET_INVALID_ARGUMENT;
+  }
+
+  event->impl->rmw_handle =
+    rmw_create_subscription_event(subscription->impl->rmw_handle, rmw_event_type);
+
+  return ret;
+}
+
+rcl_ret_t
+rcl_client_event_init(
+  rcl_event_t * event,
+  const rcl_client_t * client,
+  const rcl_client_options_t * options,
+  const rcl_client_event_type_t event_type)
+{
+  rcl_ret_t ret = RCL_RET_OK;
+
+  // Check options and allocator first, so allocator can be used with errors.
+  RCL_CHECK_ARGUMENT_FOR_NULL(options, RCL_RET_INVALID_ARGUMENT);
+  rcl_allocator_t * allocator = (rcl_allocator_t *)&options->allocator;
+  RCL_CHECK_ALLOCATOR_WITH_MSG(allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+
+  // Allocate space for the implementation struct.
+  event->impl = (rcl_event_impl_t *) allocator->allocate(
+    sizeof(rcl_event_impl_t), allocator->state);
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    event->impl, "allocating memory failed", ret = RCL_RET_BAD_ALLOC; return ret);
+
+  event->impl->rmw_handle = rmw_create_client_event(client->impl->rmw_handle);
+
+  return ret;
+}
+
+rcl_ret_t
+rcl_service_event_init(
+  rcl_event_t * event,
+  const rcl_service_t * service,
+  const rcl_service_options_t * options,
+  const rcl_service_event_type_t event_type)
+{
+  rcl_ret_t ret = RCL_RET_OK;
+
+  // Check options and allocator first, so allocator can be used with errors.
+  RCL_CHECK_ARGUMENT_FOR_NULL(options, RCL_RET_INVALID_ARGUMENT);
+  rcl_allocator_t * allocator = (rcl_allocator_t *)&options->allocator;
+  RCL_CHECK_ALLOCATOR_WITH_MSG(allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+
+  // Allocate space for the implementation struct.
+  event->impl = (rcl_event_impl_t *) allocator->allocate(
+    sizeof(rcl_event_impl_t), allocator->state);
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    event->impl, "allocating memory failed", ret = RCL_RET_BAD_ALLOC; return ret);
+
+  event->impl->rmw_handle = rmw_create_service_event(service->impl->rmw_handle);
+
+  return ret;
+}
+
+rcl_ret_t
+rcl_take_event(
+  const rcl_event_t * event,
+  void * event_status)
+{
+  bool taken;
+  RCL_CHECK_ARGUMENT_FOR_NULL(event, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(event_status, RCL_RET_INVALID_ARGUMENT);
+  rmw_ret_t ret = rmw_take_event(event->impl->rmw_handle, event_status, &taken);
```

Comment:
Seems a little odd that we're diverging `rcl_take_event` and `rmw_take_event` in terms of having the extra parameter to determine if something was actually taken. Why can't they both work the same.

Reply:
I am for removing the `taken` parameter.

Reply:
Will remove the `taken` parameter.

Reply:
On second thought, it looks like the other `rmw_take()` functions also have this extra `taken` boolean. We should probably leave it in there to be consistent.